### DISCUSSION TITLE
Keycloak `Provider` dependency graph?

### DISCUSSION THREAD

**Discussion by xgp** [keycloak/keycloak#32278] [Q&A]
I've been working on some documentation about Keycloak extensions and implementations/replacements of SPIs. One of the things that would be useful to know is a dependency graph of all of the `Provider`s. Does anyone know if such a document exists, or if there is a programmatic way to generate it?

**Comment by danielFesenmeyer** [comment]
Just wondering what you mean with "dependency graph"? Do you mean dependencies regarding startup order or usage of other providers? But wouldn't the latter be "Provider implementation"-specific?

**Comment by xgp** [comment]
Both. And yes, the latter would be implementation-specific. One of the reasons I’m interested in what providers are used by others is so I can easily answer the question, “If I replace or remove provider X, what provides are effected?”
